### ISSUE TITLE
Log levels

### ISSUE THREAD

**Issue by roshanrags** [akolov/swift-logging#2]
Proposal: [SWIFTLOG-0003](https://github.com/akolov/swift-logging/blob/master/proposals/0003-logger-api.md)

The log levels used in the reference projects are usually Verbose, Debug, Info, Warn and Error. Any reasons for getting rid of Verbose and adding a Critical level on top of Error?

EDIT: It would also be beneficial to document what the various levels signify(something like [this](https://github.com/nlog/nlog/wiki/Log-levels))

**Comment by sashka** [comment]
From my point of view Verbose is a shortcut for `please log messages from all levels`, therefore it's not a level itself. Meanwhile it may be useful as a special predefined case.

**Comment by akolov** [comment]
I just partially reused log levels available in ASL and python-logging and we might get rid of alert and critical ones. I do like emergency level though.
As for verbose, I agree with @sashka that it sounds more like "show me everything". But some libraries define one additional level after debug — the `trace` level. But in situations that I've encountered I didn't see the need for additional debugging level. Can you give us some usage and maybe real world examples?

**Comment by akolov** [comment]
For the significance it'd make sense to assign them when we stabilize/decide which levels we'll have. Right now I'm open to any ideas on that.

**Comment by roshanrags** [comment]
Yes, the `Verbose`(or `trace`) level usually serves as a trace on the program. You record entering/exiting functions, etc there. It is more about program flow whereas `Debug` is more about data(you log the web request that was sent or the query executed on a database).

Personally, I haven't used `trace` so I don't really know whether the above distinction is significant enough to warrant a separate level. I've just seen it around in a few libraries.

One benefit that I can see is since it has to do with program flow, it provides an avenue for log messages to be inserted automatically(say, the compiler can insert function entry/exit messages).

**Comment by akolov** [comment]
Yeah, data and method trace separation makes sense. But in this case I would not include these calls in optimized builds as trace specifically is very expensive.

**Comment by roshanrags** [comment]
+1 on that.

**Comment by skreutzberger** [comment]
Coming from Python, PHP and Nodejs (like many Swift Newbies will do) I recommend changing `Warn` to `Warning`. It is an issue I often run into because all levels are nouns just Warn is not a noun but a verb. In best case both levels are possible and have the same level priority.

Adding more levels is fine, too but I think an error is the worst case that happen in an app so `Error` should be enough I guess. I worked with many developers in the past and most are anyway just using `Debug`, `Info` and `Error`.

**Comment by akolov** [comment]
@skreutzberger, isn't it already `Warning` in the draft?

**Comment by akolov** [comment]
We probably don't need more levels above `Error` since anything that doesn't make application break: crash, segfault or put it into critical state could be sent as a `Warning`. I'd love to hear everyone's opinion though, since it's a fundamental topic.

**Comment by akolov** [comment]
So, revisiting this, how would we work around end-points having more log levels than defined in logging.core? 
E.g. if we were to remove `emergency`, `alert` and `critical` how would we deal with ASL, which supports them?

**Comment by skreutzberger** [comment]
For the sake of simplicity I recommend ignoring it. ASL is not widely used anymore AFAIR and maybe we better orient on standards that are used among multiple programming languages not looking back what Apple did in the past.

**Comment by jeremy-pereira** [comment]
I speak as a programmer of 30 years who has never used any level above `error` in an application, but I have to ask what the cost is of keeping the higher log levels? They will just be extra enum cases, right?

The higher levels are for more system like things, for instance the OS X syslog man page claims that `emergency` is a synonym for "panic". However, who said that Swift will never be used for doing system like things? For instance, it seems quite reasonable to me that you could write an application in Swift that monitors disk usage and that might want to issue an `alert` message when a disk becomes 95% full and a `critical` message when a disk is full.

**Comment by skreutzberger** [comment]
Surely / hopefully Swift will be used for system-level apps :)

Why not use `Warning` for `Alert` and `Error` for `Critical`? There is no extra cost in having more levels besides uncertainty on the user side which may actually be a big cost in terms of experience.

**Comment by akolov** [comment]
What about this (shamefully stolen from log4j documentation):
- `warning` — for potentially harmful situations
- `error` — error events that might still allow the application to continue running
- `critical` — events that will presumably lead the application to abort

**Comment by sashka** [comment]
From my perspective it'd be great to keep the list of levels minimal and self descriptive. We may add more levels to the library later, or let developers to add their own level if they need.

From my own experience I've never used any levels besides these:
`debug` — to show something necessary only during development and debugging;
`info` — to inform user about current state of application itself;
`warning` — to let user know that something went slightly not the way it was expected to go;
`error` — something bad happened, I want user to know more about this, and, maybe to send me a trace.

Critical/fatal means for me, that an app is going to be aborted soon after the error. I may be completely wrong, but I'd rather add a flag `fatal` to the error information that should be passed to `error` method.

**Comment by akolov** [comment]
Okay, let's keep it simple: I will make a new proposal specifically addressing log levels with proposed variants and reasoning behind and we will vote for them.

**Comment by jeremy-pereira** [comment]
@skreutzberger   

> Why not use Warning for Alert and Error for Critical? 

Because then you lose those levels for application level events. 

> There is no extra cost in having more levels besides uncertainty on the user side which may actually be a big cost in terms of experience.

I don't think it will be a big cost in terms of experience. Logging systems have had these levels (or equivalent) for many years without thousands of programmers getting confused. We are talking about keeping a couple of extra enum cases which will have a minimal cost in terms of learning curve. 

@sashka 

> From my perspective it'd be great to keep the list of levels minimal and self descriptive.

I think having alert and critical _is_ part of the minimal set.  You might change their names slightly or have some documentation to make it clear that they are meant for system level logging, not application level logging but they have minimal cost and expand the horizons of the module.

**Comment by spoof** [comment]
Actually I don't remember when I used `critical` level last time and I can't even recall good examples of its usage. Thus I see only one reason to include this level — to make swift-logging look like widely used logging libraries/framework in other languages. And it's a good reason.
Anyway, If we someday realize we need `critical` level, we can add it with no cost (it's expanding but not shrinking). So I vote for `error`, `warning`, `info`, `debug` to be in the core.
